stdlib/public/core/StringIndex.swift

```diff
@@ -62,10 +62,22 @@ extension String.Index {
     @inline(__always) get { return orderingValue == 0 }
   }
 
+  /// The UTF-16 code unit offset corresponding to this Index
+  public func utf16Offset<S: StringProtocol>(in s: S) -> Int {
+    return s.utf16.distance(from: s.utf16.startIndex, to: self)
+  }
+
   /// The offset into a string's code units for this index.
+  @available(swift, deprecated: 4.2, message: """
+    encodedOffset has been deprecated as most common usage is incorrect. \
+    Use utf16Offset(in:) to achieve the same behavior.
+    """)
```

Comment:
* Should the initializer *and* property be `deprecated: 4.2` or `deprecated: 5.0`?

* Should the deprecated APIs be moved to [MigrationSupport.swift](https://github.com/apple/swift/blob/master/stdlib/public/core/MigrationSupport.swift) if possible?

* A multi-line message (even with escaped line breaks) won't display properly in Xcode 10.1 (and *perhaps* in Xcode 10.2 beta).

Reply:
1. I believe that availability matches the language version and not deployment target. This should be deprecated regardless of what is passed in `-swift-version`. @jrose-apple / @moiseev what do you think?

2. I think MigrationSupport is done for obsoleted. We should probably obsolete this around 5.1 or so and move it then.

3. I'll have to try that then.

Reply:
MigrationSupport has some non-obsoleted APIs:

* [`DictionaryLiteral`](https://github.com/apple/swift/blob/0fe7e148c6f5750e9320ecae685ea0698b07c5b8/stdlib/public/core/MigrationSupport.swift#L180-L181)

* [`UTF8`, `UTF16`, `UTF32`, `UnicodeScalar`](https://github.com/apple/swift/blob/0fe7e148c6f5750e9320ecae685ea0698b07c5b8/stdlib/public/core/MigrationSupport.swift#L248-L255) (which are also non-deprecated).

* [`_boundsCheck(_:)`](https://github.com/apple/swift/blob/0fe7e148c6f5750e9320ecae685ea0698b07c5b8/stdlib/public/core/MigrationSupport.swift#L304-L309) (I'm not sure what this is doing here, it seems to be unused).

* [`index(where:)`, `index(of:)`](https://github.com/apple/swift/blob/0fe7e148c6f5750e9320ecae685ea0698b07c5b8/stdlib/public/core/MigrationSupport.swift#L557-L577)